### PR TITLE
Escrow smart contract

### PR DESCRIPTION
**Pull request by christine-sun**
### Summary
Added staking crypto functionality to frontend in requester/create-task with @jennsun . Note that backend is at https://github.com/jennsun/escrow-smart-contract

After a Requester creates a task
- a new Escrow contract is created for this task (spawning from the EscrowContractFactory)
- the new contract address is retrieved
image

- the user grants permission for the contract to address their WMATIC
- the contract receives cryptoAllocated amount of WMATIC

Updated cryptoAllocated to be a string for compatibility with the BigNumber object from ethers.js

### ToDo
- [ ] Upload the newContractAddress with the rest of the Task information to MoralisDB @nicholaspad 
- [ ] Consolidate Metamask popups to improve user experience

### Considerations
Used WMATIC due to its ERC20 compatibility with the smart contracts. However, does this add too much friction to the user experience in requiring users to hold and exchange WMATIC?

### CODE REVIEW

**Comment by jennsun** [comment]
Video demonstrating flow of creating task in action: 3 metamask popups corresponding to creation of smart contract, using WMATIC, and funding the contract.

https://user-images.githubusercontent.com/26675662/163757258-579b0916-ad98-41a6-939a-45f79edc4c9b.mp4

File: package.json
Comment (nicholaspad) [inline package.json:19]: `package-lock.json` again has a 10k line diff, `npm i` says that it's because you're using an old version of npm. I'll fix on my end and push.

File: pages/requester/create-task.tsx
Comment (nicholaspad) [inline pages/requester/create-task.tsx:32]: I would move these to our `.env.local`. Add these lines:

```
NEXT_PUBLIC_ESCROW_FACTORY_ADDRESS=0x5dbd6085d4bDCb57eBF5E4b2817D6e20DdEE136b
NEXT_PUBLIC_MATIC_TOKEN_ADDRESS=0x9c3C9283D3e44854697Cd22D3Faa240Cfb032889
```

and reference them in the code with `process.env.VAR_NAME`.

Make sure not to push your `.env.local`.

File: pages/requester/create-task.tsx
Comment (nicholaspad) [inline pages/requester/create-task.tsx:35]: This variable is unused - if this is intended, I would remove this line.

File: pages/requester/create-task.tsx
Comment (nicholaspad) [inline pages/requester/create-task.tsx:34]: Referring to `user?` here is unpredictable. I see that you use `requesterAddress` on line 140 - I would put `user.get(...)` (no question mark) directly on that line, and in the surrounding conditional, I would change it to `if (ethereum && user) {...}`.

File: pages/requester/create-task.tsx
Comment (nicholaspad) [inline pages/requester/create-task.tsx:145]: Got it! I'll handle this once this PR is merged.

@nicholaspad

File: pages/requester/create-task.tsx
Comment (nicholaspad) [inline pages/requester/create-task.tsx:102]: Can you also print the error?

```javascript
alert(`There was an error creating this task: ${stakeCryptoError}`)
```

File: src/utils/abi/ERC20Token.json
Comment (nicholaspad) [inline src/utils/abi/ERC20Token.json:1]: I don't know how to review this file so I'll leave it to you to double check it!

**Comment by nicholaspad** [comment]
> Video demonstrating flow of creating task in action: 3 metamask popups corresponding to creation of smart contract, using WMATIC, and funding the contract.
> 
>  2022-04-18.01-00-36.mp4

Also, can you say a little more about what each of the 3 popups does? Is the smart contract specifically associated with the task? What does "using WMATIC" mean?

File: pages/transfer-crypto.tsx
Comment (christine-sun) [reply pages/transfer-crypto.tsx:13]: We started this page to start testing withdrawing ETH. This is not related to staking ETH in the contract for the current Task and will be removed once the contract address is retrieved for each task and Taskers can withdraw ETH

**Comment by christine-sun** [comment]
Yes, the smart contract is specifically for this task. Edited @jennsun 's comment to be clearer about the Metamask popups, but will work on consolidating them to 1 to reduce user friction

File: pages/requester/create-task.tsx
Comment (nicholaspad) [reply pages/requester/create-task.tsx:34]: Also, `@ts-expect-error` should be used rarely, if at all. For example, if there's a 100% unavoidable type error from an external library that has a bug or issue related to its types, then it might be ok.

File: pages/requester/create-task.tsx
Comment (nicholaspad) [inline pages/requester/create-task.tsx:104]: The embedded variable should be `stakeCryptoErr`. I'll fix and push.

File: pages/requester/create-task.tsx
Comment (nicholaspad) [inline pages/requester/create-task.tsx:26]: These imports are unused - I'll remove them on my end.

**Review by nicholaspad — APPROVED** [review]
Looks good. I pushed a commit to get rid of unused imports and fix a variable name typo. Also merged from main.

**Pull request merged**
